### ISSUE TITLE
ReplicaSet names no longer match pod-template-hash label

### ISSUE THREAD

**Issue by abstrctn** [kubernetes/kubernetes#55346]
/kind bug

**What happened**:
Generated names of ReplicaSets follow a new (more consistent) pattern introduced with #51538. However, this means that the names of ReplicaSets no longer match the `pod-template-hash` label added to ReplicaSets and their Pods.

From the [Workloads documentation](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#creating-a-deployment):
> Notice that the name of the ReplicaSet is always formatted as `[DEPLOYMENT-NAME]-[POD-TEMPLATE-HASH-VALUE]`. The hash value is automatically generated when the Deployment is created.

**What you expected to happen**:
I expected to be able to identify a ReplicaSet's owner Deployment by removing the `pod-template-hash` from its name, which the documentation suggests is possible.

**How to reproduce it (as minimally and precisely as possible)**:
Create a new deployment in a Kubernetes 1.8.x cluster: `kubectl run my-nginx --image=nginx --replicas=2 --port=80`. Compare the resulting ReplicaSet's `.metadata.name` to its `.metadata.labels.pod-template-hash`.

**Anything else we need to know?**:
It seems like the pod-template-hash label now serves a distinct purpose from the generated name, and the documentation should probably be updated to not suggest they're linked. The proper way to identify parent resources seems to be to use the `.metadata.ownerReferences` array on ReplicaSets and Pods. Mentioning this is the docs could be useful.

**Environment**:
- Kubernetes version (use `kubectl version`): 1.8.1-gke.1
- Cloud provider or hardware configuration: GKE

**Comment by rishi-anand** [comment]
![image](https://user-images.githubusercontent.com/9067666/32978425-22f45106-cc68-11e7-93fc-8ef907c88ff8.png)

I found this PR. Here they are getting the hash and then they [safe encode](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/rand/rand.go#L112-L120) it. 

```
// SafeEncodeString encodes s using the same characters as rand.String. This reduces the chances of bad words and
// ensures that strings generated from hash functions appear consistent throughout the API.
func SafeEncodeString(s string) string {
	r := make([]byte, len(s))
	for i, b := range []rune(s) {
		r[i] = alphanums[(int(b) % len(alphanums))]
	}
	return string(r)
}
```

**Comment by rishi-anand** [comment]
In my case, pod-template-hash is '1250338284' and replica set name is 'nginx-deployment-569477d6d8' for deployment 'nginx-deployment'.

**Comment by janetkuo** [comment]
This is a regression and the fix was just merged: #66094
